src/Grpc.Net.Client/Balancer/SubchannelsLoadBalancer.cs

```diff
@@ -0,0 +1,296 @@
+        public override void UpdateChannelState(ChannelState state)
+        {
+            if (state.Status.StatusCode != StatusCode.OK)
+            {
+                ResolverError(state.Status);
+                return;
+            }
+            if (state.Addresses == null || state.Addresses.Count == 0)
+            {
+                ResolverError(new Status(StatusCode.Unavailable, "Resolver returned no addresses."));
+                return;
+            }
+
+            var allUpdatedSubchannels = new List<AddressSubchannel>();
+            var newSubchannels = new List<Subchannel>();
+            var currentSubchannels = _addressSubchannels.ToList();
+
+            foreach (var address in state.Addresses)
+            {
+                var i = FindSubchannelByAddress(currentSubchannels, address);
+
+                AddressSubchannel newSubConnection;
+                if (i != null)
+                {
+                    newSubConnection = currentSubchannels[i.GetValueOrDefault()];
+                    currentSubchannels.RemoveAt(i.GetValueOrDefault());
+                }
+                else
+                {
+                    var c = Controller.CreateSubchannel(new SubchannelOptions(new[] { address }));
+                    c.StateChanged += UpdateSubchannelState;
+
+                    newSubchannels.Add(c);
+                    newSubConnection = new AddressSubchannel(c, address);
+                }
+
+                allUpdatedSubchannels.Add(newSubConnection);
+            }
+
+            // Any sub-connections still in this collection are no longer returned by the resolver.
+            // This can all be removed.
+            var removedSubConnections = currentSubchannels;
+
+            if (removedSubConnections.Count == 0 && newSubchannels.Count == 0)
+            {
+                _logger.LogTrace("Connections unchanged.");
+                return;
+            }
+
+            foreach (var removedSubConnection in removedSubConnections)
+            {
+                RemoveSubchannel(removedSubConnection.Subchannel);
+            }
+
+            _addressSubchannels.Clear();
+            _addressSubchannels.AddRange(allUpdatedSubchannels);
+
+            // Start new connections after collection on balancer has been updated.
+            foreach (var c in newSubchannels)
+            {
+                c.RequestConnection();
+            }
+
+            UpdateBalancingState(state.Status);
+        }
+
+        private void UpdateBalancingState(Status status)
+        {
+            var readySubchannels = _addressSubchannels
+                .Select(s => s.Subchannel)
+                .Where(s => s.State == ConnectivityState.Ready)
+                .ToList();
+
+            if (readySubchannels.Count == 0)
+            {
+                // No READY subchannels, determine aggregate state and error status
+                var isConnecting = false;
+                ConnectivityState? aggState = null;
```

Comment:
Where is this used? It seems to be only assigned.

Reply:
Removed